WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+			uint64_t frequency = divider * afeClock;
+			// Calculate the deviation
+			uint64_t deviation = (frequency > central ? frequency - central : central - frequency) * 10000 / central;
+			DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Dev = %6llu; Central = %10llu Hz; DCO Freq = %12llu Hz; Divider = %2d.\n", deviation, central, frequency, divider);
+			
+			// Guard: Positive deviation is within the allowed range
+			if (frequency >= central && deviation >= SKL_DCO_MAX_POS_DEVIATION)
+				continue;
+			
+			// Guard: Negative deviation is within the allowed range
+			if (frequency < central && deviation >= SKL_DCO_MAX_NEG_DEVIATION)
+				continue;
+			
+			// Guard: Less than the current minimum deviation value
+			if (deviation >= context.minDeviation)
+				continue;
+			
+			// Found a better one
+			// Update the value
+			context.minDeviation = deviation;
+			context.central = central;
+			context.frequency = frequency;
+			context.divider = divider;
+			DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: FOUND: Min Dev = %8llu; Central = %10llu Hz; Freq = %12llu Hz; Divider = %d\n", deviation, central, frequency, divider);
+			
+			// Guard: Check whether the new minmimum deviation has been reduced to 0
+			if (deviation != 0)
+				continue;
+			
+			// Guard: An even divider is preferred
+			if (divider % 2 == 0) {
+				DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Found an even divider [%d] with deviation 0.\n", divider);
+				break;
+			}
+		}
+	}
+	
+	// Guard: A valid divider has been found
+	if (context.divider == 0) {
+		SYSLOG("igfx", "SC: ComputeHdmiP0P1P2() Error: Cannot find a valid divider for the given pixel clock %d Hz.\n", pixelClock);
+		return 0;
+	}
+	
+	// Calculate the p,q,k dividers
+	populateP0P1P2(&context);
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Divider = %d --> P0 = %d; P1 = %d; P2 = %d.\n", context.divider, context.pdiv, context.qdiv, context.kdiv);
+	
+	// Calculate the CRTC parameters
+	uint32_t multiplier = (uint32_t) (context.frequency / 24000000);
+	uint32_t fraction = (uint32_t) (context.frequency - multiplier * 24000000);
+	uint32_t cf15625 = (uint32_t) (context.central / 15625);
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Multiplier = %d; Fraction = %d; CF15625 = %d.\n", multiplier, fraction, cf15625);
+	// Guard: The given CRTC parameters should never be NULL
+	auto params = reinterpret_cast<CRTCParams*>(parameters);
+	if (params == nullptr) {
+		DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() Error: The given CRTC parameters should not be NULL.");
+		return 0;
+	}
+	params->pdiv = context.pdiv;
+	params->qdiv = context.qdiv;
+	params->kdiv = context.kdiv;
+	params->multiplier = multiplier;
+	params->fraction = fraction;
+	params->cf15625 = cf15625;
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: CTRC parameters have been populated successfully.");
+	return 0;
+}
+
+IGFX::LSPCON::LSPCON(void *controller, IORegistryEntry *framebuffer, void *displayPath) {
+	this->controller = controller;
```

Comment:
Maybe put the constructor to header? It will help to find where members are initialised a lot quicker.